TNL-4457: Remove python false positives

robrap opened the pull request:
## [TNL-4457](https://openedx.atlassian.net/browse/TNL-4457)

TNL-4457: Remove python false positives

Fix Python false positives with AST.
- Use AST to parse Python.
- Refactor Mako to use Python linting.
- Add optional verbose script argument.
- Skip tests.py files when linting.

As part of refactoring Mako to use Python linting code, I made the following violation changes:
- Drop mako_html_requires_text in favor of python_requires_html_or_text
- Retire mako_html_alone
- Retire mako_text_redundant

### Testing
- [x] Data is properly encoded in HTML templates to avoid XSS
- [x] Unit, integration, acceptance tests as appropriate

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [ ] Code review: @efischer19 
- [ ] Code review: @nedbat 

### Post-review
- [ ] Squash commits

robrap commented:
@efischer19 @nedbat: FYI: I had to factor out a move of MakoTemplateLinter to a different branch so you could actually see the changes.  I closed the other PR in favor of this one.  Thanks.

efischer19 commented:
👍 Very detailed, good test coverage.

The quality tests are failing because of this line, https://github.com/edx/edx-platform/pull/12320/files#diff-9ab325934d5059e5d2a330f5020c8cf7R2030:
```
File "/home/jenkins/workspace/edx-platform-quality-pr/edx-platform/scripts/safe_template_linter.py", line 2030, in check_python_file_is_safe
self._check_concat_with_html(file_contents, Rules.python_concat_html, results)
AttributeError: 'PythonLinter' object has no attribute '_check_concat_with_html'
```
https://build.testeng.edx.org/job/edx-platform-quality-pr/17329/console

robrap commented:
Thanks @efischer19.  When I split the original PR into 2, I had a bad merge and there were two copies of  PythonLinter (the original and the new).  I just added a new PR that deleted the old PythonLinter.  Sorry about that.

nedbat commented on scripts/safe_template_linter.py:
```diff
@@ -1406,6 +1419,375 @@ def _is_jquery_insert_caller_safe(self, caller_line_start):
+                format_caller = node.func.value
+                is_caller_html_or_text = isinstance(format_caller, ast.Call)
+                is_caller_html_or_text = is_caller_html_or_text and isinstance(format_caller.func, ast.Name)
+                is_caller_html_or_text = is_caller_html_or_text and format_caller.func.id in ['Text', 'HTML']
```
This would be more understandable as a single expression on multiple lines, I think.

robrap replied:
Sounds good.

nedbat commented on scripts/safe_template_linter.py:
```diff
@@ -1406,6 +1419,375 @@ def _is_jquery_insert_caller_safe(self, caller_line_start):
+            visitor = HtmlStringVisitor(self.file_contents, self.results)
+            visitor.visit(node.left)
+            has_illegal_html_string = len(visitor.unsafe_html_string_nodes) > 0
+            visitor = HtmlStringVisitor(self.file_contents, self.results)
```
Why re-make the visitor?

robrap replied:
It tracks state, so either start fresh or have some sort of `reset_state()` function and the former doesn't require a new function or any tests.  Note that the visitor isn't parsing, it's just walking the tree, so hopefully this isn't a big deal either way.

robrap replied:
Note, if you still think it would be better to have a `reset_state()` and/or to call it at the top of its `visit()` function, just let me know.

robrap replied:
@nedbat ^

nedbat replied:
If you need a new one, then it's fine to make a new one.  It looked like a leftover line.

nedbat commented on scripts/safe_template_linter.py:
```diff
@@ -1464,244 +1852,137 @@ def check_python_file_is_safe(self, file_contents, results):
+        # PEP-263 Provides Regex for Declaring Encoding
+        # Example: -*- coding: <encoding name> -*-
+        # This is only allowed on the first two lines, and it must be stripped
+        # before parsing, because we have already read into Unicode and the
```
This is *so* annoying.  It's the stupidest thing, who thought it was a good idea! :(

nedbat commented on scripts/safe_template_linter.py:
```diff
@@ -1464,244 +1852,137 @@ def check_python_file_is_safe(self, file_contents, results):
-                results.violations.append(ExpressionRuleViolation(
-                    Rules.python_wrap_html, unsafe_string_expression
-                ))
+        for match in re.finditer(r'\.display_name_with_default_escaped', file_contents):
```
Seems odd to still use a regex for this, since you are visiting all the nodes anyway.

robrap replied:
@nedbat: It is odd.  This was pre-existing code, and I needed to draw the line somewhere on the refactor.  I could add `TODO` comments, but I was trying to draw the line somewhere.  The same thing could be said for `python_custom_escape` below, but that gets slightly more complicated.  I may see how quickly I could take care of 1 or both.  Ugh. ;)

nedbat commented on scripts/safe_template_linter.py:
```diff
@@ -1464,244 +1852,137 @@ def check_python_file_is_safe(self, file_contents, results):
     The linter for Mako template files.
     """
-    javaScriptLinter = JavaScriptLinter()
+    javascript_linter = JavaScriptLinter()
```
These would be better as instance attributes in __init__, no?

robrap replied:
Sure.  I'm not sure why they are out here, but probably from before __init__ and when I was struggling getting some Python class stuff working.

nedbat commented on scripts/safe_template_linter.py:
```diff
@@ -553,27 +544,39 @@ def prepare_results(self, full_path, string_lines):
+            try:
+                line = self.lines[line_number - self.start_line].encode(encoding='utf-8')
+            except (TypeError, IndexError):
+                line = "Error printing line with line_number={} and start_line={}.".format(line_number, self.start_line)
```
What can cause these errors?

robrap replied:
@nedbat: A programming error elsewhere along the way.  I can drop this now if you think it would be better not to have this around.

nedbat replied:
If this can't happen now, it would be better to get rid of the defensive code.

nedbat commented on scripts/safe_template_linter.py:
```diff
@@ -1102,6 +1114,7 @@ def __init__(self):
         super(JavaScriptLinter, self).__init__()
         self._skip_javascript_dirs = self._skip_dirs + ('i18n', 'static/coffee')
         self._skip_coffeescript_dirs = self._skip_dirs
+        self.line_comment_delim = "//"
```
This is a constant that seems like it would be right as a class attribute.

robrap replied:
Ok.

nedbat commented on scripts/safe_template_linter.py:
```diff
@@ -1406,6 +1419,375 @@ def _is_jquery_insert_caller_safe(self, caller_line_start):
+        if isinstance(node, ast.Str):
+            # Triple quotes give col_offset of -1 on the last line of the string.
+            if node.col_offset == -1:
+                triple_quote_regex = re.compile("""['"]{3}""")
```
This might get the wrong answer for nested triple quotes, is that OK?

robrap replied:
I have an idea.  I'll try something here.

robrap replied:
The following should work for almost anything unless you intentionally try to trick it.  I added tests.

nedbat commented on scripts/safe_template_linter.py:
```diff
@@ -1406,6 +1419,375 @@ def _is_jquery_insert_caller_safe(self, caller_line_start):
+            self.generic_visit(node)
+
+
+class CloseBeforeFormatVisitor(BaseVisitor):
```
This visitor doesn't do anything about "close before format".  All it does is notice if there was a .format function call.

nedbat commented on scripts/safe_template_linter.py:
```diff
@@ -1406,6 +1419,375 @@ def _is_jquery_insert_caller_safe(self, caller_line_start):
+        """
+        # Attribute(expr value, identifier attr, expr_context ctx)
+        if node.attr == 'format':
+            self.contains_format_call = True
```
This will be set to True for `x = y.format`, won't it? Will that be OK?

robrap replied:
@nedbat: I think this will just have to be ok.  
1. I imagine this isn't really causing an issue.
2. Unless you can help me see otherwise, format is an `ast.Attribute`, for which it is really easy to determine the left hand side (i.e. what is this an attribute of), but very difficult to get the right-hand side (i.e. what arguments are we passing to format()).

At this point, I plan on leaving this as-is.  Sound ok?

robrap commented on scripts/safe-commit-linter.sh:
```diff
@@ -18,6 +18,10 @@ show_help() {
     echo "                              Defaults to origin/master."
+    echo ""
+    echo "For additional help:"
+    echo "  http://edx.readthedocs.org/projects/edx-developer-guide/en/latest/conventions/safe_templates.html#safe-template-linter"
```
@nedbat: I snuck in this little change.

robrap commented:
@nedbat: I addressed your comments through combo of code changes or responses.  Let me know when all is well. Thank you!

nedbat commented:
:+1:

robrap commented:
Merge to master will happen in a separate PR: https://github.com/edx/edx-platform/pull/12352
